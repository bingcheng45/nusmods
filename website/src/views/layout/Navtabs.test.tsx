import { render } from '@testing-library/react';
import configureStore from 'bootstrapping/configure-store';
import produce from 'immer';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import reducers from 'reducers';
import { initAction } from 'test-utils/redux';

import Navtabs from './Navtabs';

const relevantStoreContents = {
  app: { activeSemester: 1 },
  settings: { beta: false },
};

const initialState = reducers(undefined, initAction());

function make(storeOverrides: Partial<typeof relevantStoreContents> = {}) {
  const { store } = configureStore(
    produce(initialState, (draft) => {
      draft.app.activeSemester =
        storeOverrides.app?.activeSemester ?? relevantStoreContents.app.activeSemester;
      draft.settings.beta = storeOverrides.settings?.beta ?? relevantStoreContents.settings.beta;
    }),
  );
  return render(
    <MemoryRouter>
      <Provider store={store}>
        <Navtabs />,
      </Provider>
    </MemoryRouter>,
  );
}

describe(Navtabs, () => {
  test('should render into nav element', () => {
    const { container } = make();
    expect(Array.from(container.getElementsByTagName('a')).map((elem) => elem.textContent))
      .toMatchInlineSnapshot(`
      Array [
        "Today",
        "Timetable",
        "Modules",
        "Venues",
        "Settings",
        "Contribute",
        "Whispers",
      ]
    `);
    expect(container).toMatchSnapshot();
  });

  test('should show beta tabs if beta is true', () => {
    const { container } = make({ settings: { beta: true } });
    expect(Array.from(container.getElementsByTagName('a')).map((elem) => elem.textContent))
      .toMatchInlineSnapshot(`
      Array [
        "Today",
        "Timetable",
        "Modules",
        "Venues",
        "Planner",
        "Settings",
        "Contribute",
        "Whispers",
      ]
    `);
    expect(container).toMatchSnapshot();
  });
});
