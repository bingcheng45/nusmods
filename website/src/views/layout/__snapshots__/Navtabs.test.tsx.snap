// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navtabs should render into nav element 2`] = `
<div>
  <nav
    class="nav"
  >
    <a
      class="link"
      href="/today"
    >
      <div
        data-testid="react-feather Clock icon"
      />
      <span
        class="title"
      >
        Today
      </span>
    </a>
    <a
      class="link"
      href="/timetable/sem-1"
    >
      <div
        data-testid="react-feather Calendar icon"
      />
      <span
        class="title"
      >
        Timetable
      </span>
    </a>
    <a
      class="link"
      href="/modules?sem[0]=1&sem[1]=2&sem[2]=3&sem[3]=4"
    >
      <div
        data-testid="react-feather BookOpen icon"
      />
      <span
        class="title"
      >
        Modules
      </span>
    </a>
    <a
      class="link"
      href="/venues"
    >
      <div
        data-testid="react-feather Map icon"
      />
      <span
        class="title"
      >
        Venues
      </span>
    </a>
    <a
      class="link"
      href="/settings"
    >
      <div
        data-testid="react-feather Settings icon"
      />
      <span
        class="title"
      >
        Settings
      </span>
    </a>
    <a
      class="link hiddenOnMobile"
      href="/contribute"
    >
      <div
        data-testid="react-feather Star icon"
      />
      <span
        class="title"
      >
        Contribute
      </span>
    </a>
    <div
      class="divider"
    />
    <a
      class="link hiddenOnMobile"
      href="https://nuswhispers.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        data-testid="react-feather Heart icon"
      />
      <span
        class="title"
      >
        Whispers
      </span>
    </a>
  </nav>
  ,
</div>
`;

exports[`Navtabs should show beta tabs if beta is true 2`] = `
<div>
  <nav
    class="nav"
  >
    <a
      class="link"
      href="/today"
    >
      <div
        data-testid="react-feather Clock icon"
      />
      <span
        class="title"
      >
        Today
      </span>
    </a>
    <a
      class="link"
      href="/timetable/sem-1"
    >
      <div
        data-testid="react-feather Calendar icon"
      />
      <span
        class="title"
      >
        Timetable
      </span>
    </a>
    <a
      class="link"
      href="/modules?sem[0]=1&sem[1]=2&sem[2]=3&sem[3]=4"
    >
      <div
        data-testid="react-feather BookOpen icon"
      />
      <span
        class="title"
      >
        Modules
      </span>
    </a>
    <a
      class="link"
      href="/venues"
    >
      <div
        data-testid="react-feather Map icon"
      />
      <span
        class="title"
      >
        Venues
      </span>
    </a>
    <a
      class="link hiddenOnMobile"
      href="/planner"
    >
      <div
        data-testid="react-feather Trello icon"
      />
      <span
        class="title"
      >
        Planner
      </span>
    </a>
    <a
      class="link"
      href="/settings"
    >
      <div
        data-testid="react-feather Settings icon"
      />
      <span
        class="title"
      >
        Settings
      </span>
    </a>
    <a
      class="link hiddenOnMobile"
      href="/contribute"
    >
      <div
        data-testid="react-feather Star icon"
      />
      <span
        class="title"
      >
        Contribute
      </span>
    </a>
    <div
      class="divider"
    />
    <a
      class="link hiddenOnMobile"
      href="https://nuswhispers.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        data-testid="react-feather Heart icon"
      />
      <span
        class="title"
      >
        Whispers
      </span>
    </a>
  </nav>
  ,
</div>
`;
